### PR TITLE
Create preprocessing docs RibaMetaMod

### PR DESCRIPTION
**Pull request by HendrikKok**

### CODE REVIEW

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+Optionally the irrigation from surface water can be coupled to the [UserDemand](https://deltares.github.io/Ribasim/reference/node/user-demand.html) in Ribasim. In that case the sprinkling realised in MetaSWAP, depends on the water availability for the [UserDemand](https://deltares.github.io/Ribasim/reference/node/user-demand.html) in Ribasim. This coupling requires the following [MetaSWAP](https://deltares.github.io/imod-python/api/msw.html) packages:
+
+* [Sprinkling](https://deltares.github.io/imod-python/api/generated/msw/imod.msw.Sprinkling.html) to define the maximum irrigation rates.
+* [LanduseOptions](https://deltares.github.io/imod-python/api/generated/msw/imod.msw.LanduseOptions.html) to specify the irrigation regime.
+
+This coupling has the following requirements on the Ribasim model:
+
+* There should be [User_demand](https://deltares.github.io/Ribasim/reference/python/Model.html#ribasim.Model.user_demand) defined with one priority per coupled water user. For this priority there should be a rate defined > 0.0
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:84]: Whereas in general water users in Ribasim can have non-zero demands set at multiple priorities,
for water users this demand can only be positive at one priority and should be zero for others.

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+If [Allocation](https://deltares.github.io/Ribasim/reference/python/Allocation.html#ribasim.Allocation) is active in the Ribasim model, The demand-realisation cyle is based on the user priority and the water availibility. When it's inactive its based on water availibility alone.   
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:86]: If Allocation  ... alone -> should be in the technical doc

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+## Coupling and model extents
+* The start and end times of the Ribasim, MetaSWAP and MODFLOW 6 simulations must align.
+  `primod` will raise an error otherwise.
+* Spatial extents of the models need not coincide:
+  + Part of the Ribasim basins may be located outside of the MODFLOW 6 and MetaSWAP simulation window. Uncoupled basins will proceed with the regular drainage and irrigation terms define in the model.
+  + Not every MODFLOW 6 River and Drainage boundary needs to be linked with
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:93]: general:"
omit "will"
may -> can

terms define -> terms defined
simulation window ->  domain 
linked with Ribasim -> completely coupled to Ribasim 
SVAT or svat ?? choose one

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  Ribasim. Boundaries outside of any basin polygon will simply use the regular
+  file input.
+  + Not every MetaSWAP SVAT need to be linked with Ribasim. For uncoupled SVATS, runoff is not routed via the surface water and irrigation from surface water is unlimited. 
+  + The MetaSWAP model does not need to be active in the complete MODFLOW 6 simulation window. Uncoupled MODFLOW 6 nodes can have an aditional [RCH](https://deltares.github.io/imod-python/api/generated/msw/imod.mf6.Recharge.html) and [EVT](https://deltares.github.io/imod-python/api/generated/mf6/imod.mf6.Evapotranspiration.html) package.
+
+
+## Abbreviated examples of coupling 
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:101]: what is an ABBREVIATED example ?

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+The following abbreviated example for coupling Runoff:
+
+* Reads a Ribasim model
+* Takes a MetaSWAP model
+* Reads a MODFLOW 6 simulation
+* Reads a basin definition associated with the Ribasim model
+* Defines a driver coupling: which river and drainage packages are coupled
+* Sets up a coupled model
+* Writes the coupled model
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:111]: example for coupling Runoff -> example of coupling runoff
Takes -> Reads or the three reads on a line: Reads a Ribasim, MetaSWAP and MODFLOW 6 model

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+ribametamod_model.write(
+   directory="coupled-ribametamod-simulation",
+   modflow6_dll=r"c:\bin\imod_coupler\modflow6\libmf6.dll",
+   ribasim_dll=r"c:\bin\imod_coupler\ribasim\bin\libribasim.dll",
+   ribasim_dll_dependency=r"c:\bin\imod_coupler\ribasim\bin",
+)
+```
+
+For including the optional coupling of irrigation from surface water, the folowing adition can be done:
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:158]: adition can be done -> addition is required
State that the driver_coupling_ribameta replaces the one in the upper part, but now with an extra basin definition for the water users

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+driver_coupling_ribameta = primod.RibaMetaDriverCoupling(
+    ribasim_basin_definition=basin_definition,
+    ribasim_user_demand_definition=water_users_definition,
+)
+
+# proceed as previous example
+
+```
+
+## Exchange derivation
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:173]: Exchange derivation -> Mapping definition

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+The exchanges are derived based on spatial overlap of the models. In general all couplings with ribasim are based on a spatial overlap of the basin and/or water user definition polygons. The coupling between MODFLOW 6 6 and MetaSWAP is based on overlapping grid based coordinates. The coupling based on polygon defninitions is done in the folowing way:
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:175]: The mapping between MODFLOW 6 and MetaSWAP is based on matching grid-based coordinates.
The mapping between either MODFLOW 6 or MetaSWAP and Ribasim are based on a spatial overlap of a 
Ribasim basin or water definition polygon. This polygon is provided as a geopandas.GeoDataFrame.
Mapping proceeds as follows:

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+* Rasterize the basin and/or Water users definition polygons (provided as a
+  `geopandas.GeoDataFrame`) to the MODFLOW 6 or MetaSWAP model grid.
+* Overlay the grid-based variable on the rasterized basin definition,
+  and derive for each gridcel the basin index.
+* Identify the indices of the coupled MODFLOW 6 or MetaSWAP gridcel.
+* Store the basin indices and gridcel indices in a table.
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:182]: reword the first two bullets:
* Rasterize de basin index to the M&M model grid, based on the definition polygon
* From the raster, obtain for each M&M node/svat the corresponding basin index

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+Ribasim and MODFLOW
+: For the flux exchange, the Overlay is based on the [river](https://deltares.github.io/imod-python/api/generated/mf6/imod.mf6.River.html) conductance. The indices are based on the package level. For the stage exchange, the mapping is based on the x and y locations of the [subgrid](https://deltares.github.io/Ribasim/reference/node/basin.html#subgrid) elements. For every [river](https://deltares.github.io/imod-python/api/generated/mf6/imod.mf6.River.html) conductance gridcell the neirest [subgrid](https://deltares.github.io/Ribasim/reference/node/basin.html#subgrid) element is found. 
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:185]: the Overlay -> the mapping (see further in that sentence) 
Overlay -> mapping
all SVATS where -> all svats for which 
grid based -> grid-based

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+
+## Modifications to the models
+
+Ribasim
+: The `primod.RibaMod` class makes the following alteration to the Ribasim input
+before writing the Ribasim model: 
+
+* for basins that are coupled to MODFLOW 6,
+the infiltration and drainage columns are set to `NaN` / `Null` (nodata) in
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:200]: remove "that are"
"the infilt...columns" -> "infiltration and drainage"
"This ensures ..." belongs to the indented block above
Conceptually... half ->
Coupled basins should entirely fall within the MODFLOW 6 domain.
The portion outside the MODFLOW 6 domain is excluded from infiltration and drainage.

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+This ensures that Ribasim does not overwrite the exchange flows while running
+coupled with MODFLOW 6.
+
+Conceptually, it also means that when a basin is coupled, it should generally
+located inside of the MODFLOW 6 model; after all, when half of the basin is
+located outside of the MODFLOW 6 model, it will not receive drainage or lose
+water to infiltration in that half.
+
+* For all [User_demand](https://deltares.github.io/Ribasim/reference/python/Model.html#ribasim.Model.user_demand) coupled to MetaSWAP, the rate is set to zero after initialisation of the iMOD-coupler. The rate > 0 is needed to find the coupled indices from the internal 
+[User_demand](https://deltares.github.io/Ribasim/reference/python/Model.html#ribasim.Model.user_demand) array.
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:212]: For all ...User_demand array .... is this relevant to preprocessing?

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+
+Modflow 6
+: For all actively coupled RIV packages an API-package is added to the simulation. This package is used to return the correction flux. When making a water-balance from the MODFLOW 6 output, this package output should be summed with the RIV output 
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:216]: To return the correction flux -> To apply/effect the flux correction ? 
When making a water balance -> In a water balance
summed with -> added to

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+MetaSWAP
+: curently no modifications are made in the MetaSWAP model
+
+
```
Comment (rleander) [inline docs/coupler_ribametamod_preprocessing.qmd:220]: curently -> Currently,

**Review by rleander — CHANGES_REQUESTED** [review]
minor text suggestions

File: docs/coupler_ribametamod_preprocessing.qmd [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+The [Modflow6Simulation](https://deltares.github.io/imod-python/api/generated/mf6/imod.mf6.Modflow6Simulation.html) and Ribasim [Model](https://deltares.github.io/Ribasim/python/reference/#model) can be initialised from a TOML file with associated data. The [MetaSwapModel](https://deltares.github.io/imod-python/api/generated/msw/imod.msw.MetaSwapModel.html) 
+does not support this yet. Initialising this model can be scripted. 
+
+
+## Coupling and model requirements
+The [RibaMetaMod](primod_api/RibaMetaMod.html#primod.RibaMetaMod) class includes three seperate couplings.
+
+### MODFLOW 6 and MetaSWAP
+This coupling is equal to the [MetaMod](primod_api/MetaMod.html#primod.MetaMod) driver. This coupling requires the following [MODFLOW](https://deltares.github.io/imod-python/api/mf6.html) packages:
+
```
[the remark this answers is not in the document]
Comment (HendrikKok) [reply docs/coupler_ribametamod_preprocessing.qmd:39]: done

File: docs/coupler_ribametamod_preprocessing.qmd (hunk 10 above)
Comment (HendrikKok) [reply docs/coupler_ribametamod_preprocessing.qmd:185]: I kept the overlay, because its referencing to line 179. SVATS are caps since its an abbreviation
